hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DataStreamer.java

```diff
@@ -414,6 +414,10 @@ synchronized void markFirstNodeIfNotMarked() {
     }
 
     synchronized void adjustState4RestartingNode() {
+      if (restartingNodeIndex == -1) {
+        return;
+      }
+
```

Comment:
Why is this needed?
Below there is a logic, doesn't that take care of things?
```
      if (!isRestartingNode()) {
        error = ErrorType.NONE;
      }
      badNodeIndex = -1;
    }
```

 None of your tests fails without this for me

Reply:
From past, some tests were failing in absence of this check(I lost track of that PR), I started a new PR to check the exact failures https://github.com/apache/hadoop/pull/6605/files

Reply:
Thanx, I checked the PR, the failure is due to ArrayIndexBound in our new code
```
Caused by: java.lang.ArrayIndexOutOfBoundsException: -1
 at org.apache.hadoop.hdfs.DataStreamer.setupPipelineForCreate(DataStreamer.java:1848)
 at org.apache.hadoop.hdfs.DataStreamer.run(DataStreamer.java:751)
```

Because we need the badNode & we are extracting like:
```
final DatanodeInfo badNode = nodes[errorState.getBadNodeIndex()];
```
The badNodeIndex is reset to -1, post this check, but the node is stored in the ``failed`` arraylist, can you extract the badNode from the ``failed`` ArrayList rather than tweaking this logic?

Maybe something like this
```
        final DatanodeInfo badNode = Iterables.getLast(failed);
```

Reply:
https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6605/2/testReport/ Still got some failures with Iterables approach

Reply:
Thanks for your commit, resolving this